### ISSUE TITLE
Fix Dispatch return type in TypeScript

### ISSUE THREAD

**Issue by laszlopandy** [redux-loop/redux-loop#237]
Original discussion here:
https://github.com/redux-loop/redux-loop/pull/222/files#r314969181

Redux-loop's dispatch returns a `Promise<Action>` instead of an `Action` like redux does. This is because we need to run the commands returned by the reducer. 

What is not clear to me:
1. Why does redux's dispatch return an `Action`? Is there anything which uses this return value?
2. Why does redux-loop's dispatch need to wait for the promises from `executeCmd` to resolve? (ie. why not just return without waiting for the promises to settle?)

**Comment by laszlopandy** [comment]
FYI @bdwain @domarmstrong @nweber-gh

**Comment by lorefnon** [comment]
The value returned by dispatch is primarily for use by consumer components (or whoever is dispatching). It is convenient in many scenarios: 

When we are creating connected components using Hoc (old-school API), the dispatcher component will typically not have access to the actual action dispatched. 

```typescript
const ConnectedProductForm = connect(
    function mapStateToProps(state) { ... },
    function mapDispatchToProps(dispatch) { 
        return {
           createProduct: (id) => dispatch(productAction({
               id, 
               dateTime: new Date()
           }))
        }
    }
)(ProductForm);

const ProductForm = ({createProduct}) => {
    return <form onSubmit={() => {
        // Component only has access to createProduct and 
        // not the action being dispatched
        const productPayload = createProduct();
        // productPayload now contains what was dispatched, and we can 
        // make use of the additional data that mapDispatchToProps (or the action creator) may
        // have added
    }}> ... </form>;
}
```

We could have alternatively injected just the dispatch along with the action creators, but this is slightly more convenient. 

This becomes more convenient when middlewares return value, as it enables the middleware to communicate forward to the store as well as backward to the component dispatching the action. 

Possibly the simplest example of this is the thunk middleware, which just returns whatever the thunk returns. So this enables the thunk to communicate back to the component (through the thunk middleware). 

Another is promise middleware, which returns another promise which resolves after the postSuccess/postFailure actions have been dispatched, so the component waiting on returned promise knows that when the returned promise has resolved, then success/failure action has already been handled by corresponding reducers and state tree at that point has been updated to accommodate them. 

In case of redux-loop, the component waiting on the returned promise knows that when the returned promise is resolved, the loop spawned by this action, as well as any loops spawned by any success/failure actions that are dispatched by the loop have all been resolved.

**Comment by laszlopandy** [comment]
Thanks @lorefnon. Can you take a look at https://github.com/redux-loop/redux-loop/pull/232? This PR would add functions to delay the execution of a command, and may dispatch actions after the returning promise has already resolved. 

It would be possible to make the execution of `Cmd.setTimeout` return a promise, but `Cmd.setinterval` allows a command to be executed repeatedly and so is not compatible with this idea that after the promise resolves, all actions have been dispatched. 

Of course it's possible to do the same thing already using `Cmd.run` and `Cmd.dispatch`, but that is considered an escape hatch and so it is clearly separate from the regular execution of commands.

**Comment by lorefnon** [comment]
Looking at the PR I don't get clarity on what are the kind of the use cases that `Cmd.setInterval` will address (or make easier), but we can return an [async iterator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/asyncIterator).

> it's possible to do the same thing already using Cmd.run and Cmd.dispatch

I usually await on `Cmd.dispatch(...)` in async loop functions. I didn't know they were considered "escape hatches" :)

**Comment by bdwain** [comment]
@laszlopandy i also have never found a reason to use the return value. At this point, it returns a Promise instead of the original action because it has always done that since before I started working with redux-loop and I see no reason to change it. There are some cases already where the promise will return before your nested Cmds run (like unbatched Cmd.lists), but those exceptions are listed in the docs and it was a breaking change when we changed the behavior.

@lorefnon Cmd.dispatch is not a function. It's just a symbol. do you mean `await dispatch(...)` after passing in Cmd.dispatch from your reducer? 

I think the idea with it being an escape hatch is just that the encouraged design is to use normal Cmds when possible to control how you dispatch actions and to pass params in through args in Cmd.run when you can. But sometimes that's just inconvenient or impossible, and so Cmd.getState and Cmd.dispatch were made. Technically, with Cmd.dispatch, you could always avoid using successActionCreator and failActionCreator, but that's discouraged.

**Comment by lorefnon** [comment]
> Cmd.dispatch is not a function. It's just a symbol. do you mean await dispatch(...) after passing in Cmd.dispatch from your reducer?

Yeah, sorry. I meant `dispatch` and not the symbol `Cmd.dispatch`.

**Comment by laszlopandy** [comment]
@bdwain `handleSequenceList` and `handleParallelList` both return a promise in the case of unbatched commands: `Promise.all(result.map(a => dispatch(a)))`. So the new `Cmd.setInterval` might be the first command which will dispatch after the promise returned by `executeCmd` resolves.

However I have also never needed to use this return value. And I agree that it is not a hard requirement we must respect in all cases. It is enough to document those cases in which a dispatch might come later.

**Comment by bdwain** [comment]
I think unbatched lists have the same issue as setInterval will.

**Comment by webholics** [comment]
I have an issue which seems to related to this one:

When using `ConntectedProps` from redux to type your conntected react component you are getting wrong types for bound actions. The return value is `Action` instead of `Promise<Action>`.

```
const mapStateToProps = () => ({})

const mapDispatchToProps = {
  init: coreActions.init,
}

const connector = connect(mapStateToProps, mapDispatchToProps)

type Props = ConnectedProps<typeof connector>
```

Therefore something like this in the component won't work and I didn't find any clean way to type this:

```
init().then(() => ...)
```

**Comment by bdwain** [comment]
Yeah I don't think the ConnectedProps type would work with redux-loop. It seems like it assumes the return type of dispatch remains unchanged.

I also would not rely on the return value of init anyway, as the promise may resolve before the "handling" of the action is complete, depending on how the action is handled (see above). The only way to be sure the action has been fully handled is to update the state in some way when things have finished and watch for that in mapStateToProps as some sort of `initted` prop. You can then trigger your `then` handler when `initted` switches to true. Hooks may simplify this pattern also.

**Comment by webholics** [comment]
The best solution I found for now is the following (if anyone is searching for this):

```
type MapResultToPromise<T> = T extends (...args: infer U) => infer R
  ? (...args: U) => Promise<R>
  : T
type MapDispatch<DispatchProps, Connector> = Omit<ConnectedProps<Connector>, keyof DispatchProps> &
  {
    [Property in keyof DispatchProps]: MapResultToPromise<DispatchProps[Property]>
  }
type Props = MapDispatch<typeof mapDispatchToProps, typeof connector>
```
